rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/CharacteristicLongWriteOperation.java

```diff
@@ -209,4 +214,56 @@ public Boolean call(Object emission) {
             }
         };
     }
+
+    private static Func1<Observable<? extends Throwable>, Observable<?>> errorIsRetryableAndAccordingTo(
+            final WriteOperationRetryStrategy writeOperationRetryStrategy,
+            final ByteBuffer byteBuffer,
+            final int batchSize) {
+        return new Func1<Observable<? extends Throwable>, Observable<?>>() {
+            @Override
+            public Observable<?> call(Observable<? extends Throwable> emittedOnWriteFailure) {
+                return emittedOnWriteFailure
+                        .flatMap(toLongWriteFailureOrError())
+                        .compose(writeOperationRetryStrategy)
+                        .doOnNext(repositionByteBufferForRetry());
+            }
+
+            @NonNull
+            private Func1<Throwable, Observable<WriteOperationRetryStrategy.LongWriteFailure>> toLongWriteFailureOrError() {
+                return new Func1<Throwable, Observable<WriteOperationRetryStrategy.LongWriteFailure>>() {
+                    @Override
+                    public Observable<WriteOperationRetryStrategy.LongWriteFailure> call(Throwable throwable) {
+                        if (!(throwable instanceof BleGattCharacteristicException)) {
+                            return Observable.error(throwable);
+                        }
+                        final int failedBatchNumber = calculateFailedBatchNumber(byteBuffer, batchSize);
+                        WriteOperationRetryStrategy.LongWriteFailure longWriteFailure = new WriteOperationRetryStrategy.LongWriteFailure(
+                                failedBatchNumber,
+                                (BleGattCharacteristicException) throwable
+                        );
+                        return Observable.just(longWriteFailure);
+                    }
+                };
+            }
+
+            @NonNull
+            private Action1<WriteOperationRetryStrategy.LongWriteFailure> repositionByteBufferForRetry() {
+                return new Action1<WriteOperationRetryStrategy.LongWriteFailure>() {
+                    @Override
+                    public void call(WriteOperationRetryStrategy.LongWriteFailure longWriteFailure) {
+                        final int newBufferPosition = longWriteFailure.getBatchNumber() * batchSize;
+                        byteBuffer.position(newBufferPosition);
+                    }
+                };
+            }
+
+            private int calculateFailedBatchNumber(ByteBuffer byteBuffer, int batchSize) {
+                if (byteBuffer.hasRemaining()) {
+                    return (byteBuffer.position() - batchSize) / batchSize;
```

Comment:
`(byteBuffer.position() / batchSize) - 1` seems slightly easier to understand for me :)